File: applications/MeshMovingApplication/custom_processes/impose_mesh_motion_process.cpp
//    |  /           |
//    ' /   __| _` | __|  _ \   __|
//    . \  |   (   | |   (   |\__ `
//   _|\_\_|  \__,_|\__|\___/ ____/
//                   Multi-Physics
//
//  License:        BSD License
//	                Kratos default license: kratos/license.txt
//
//  Main Authors:   Máté Kelemen
//

// Internal inlcudes
#include "impose_mesh_motion_process.h"

// Project includes
#include "custom_utilities/parametric_linear_transform.h"
#include "includes/checks.h"
#include "utilities/parallel_utilities.h"
#include "includes/mesh_moving_variables.h"


namespace Kratos
{


ImposeMeshMotionProcess::ImposeMeshMotionProcess(Model& rModel, Parameters parameters) :
    ImposeMeshMotionProcess(rModel.GetModelPart(parameters.GetValue("model_part_name").GetString()), parameters)
{
    
}


ImposeMeshMotionProcess::ImposeMeshMotionProcess(ModelPart& rModelPart, Parameters parameters) :
    Process(),
    mrModelPart(rModelPart),
    mIntervalUtility(parameters) // initialized here because IntervalUtility has no default constructor
{
    KRATOS_TRY;

    // Check model part name and validate parameters
    KRATOS_ERROR_IF(rModelPart.FullName() != parameters.GetValue("model_part_name").GetString())
    << "Expecting model part named '" << parameters.GetValue("model_part_name").GetString() << "' but got '" << rModelPart.FullName() << "' instead";

    // Process default parameters
    // 'rotation_angle' can either be numeric or an expression (string),
    // but the Parameters validator can't handle both.
    // -> default parameters must be tweaked to match the input type
    Parameters default_parameters = this->GetDefaultParameters();
    if (parameters.Has("rotation_angle") && parameters.GetValue("rotation_angle").IsString()) {
        default_parameters.RemoveValue("rotation_angle");
        default_parameters.AddValue("rotation_angle", Parameters("\"0\""));
    }

    parameters.ValidateAndAssignDefaults(default_parameters);

    // Parse general parameters
    mIntervalUtility = IntervalUtility(parameters);
    std::string rotation_definition = parameters.GetValue("rotation_definition").GetString();

    // Check whether all numeric parameters are constants
    //  - if yes -> LinearTransform
    //  - otherwise -> ParametricLinearTransform
    bool require_parametric_transform = false;

    const auto euler_angles_parameters = parameters.GetValue("euler_angles");
    const auto rotation_axis_parameters = parameters.GetValue("rotation_axis");
    const auto reference_point_parameters = parameters.GetValue("reference_point");
    const auto rotation_angle_parameters = parameters.GetValue("rotation_angle");
    const auto translation_vector_parameters = parameters.GetValue("translation_vector");

    if (rotation_angle_parameters.IsString()) {
        require_parametric_transform = true;
    }
    else {
        for (std::size_t i=0; i<3; ++i)
        {
            if (euler_angles_parameters.GetArrayItem(i).IsString()
                || rotation_axis_parameters.GetArrayItem(i).IsString()
                || reference_point_parameters.GetArrayItem(i).IsString()
                || translation_vector_parameters.GetArrayItem(i).IsString()) {
                    require_parametric_transform = true;
                    break;
            } // if any parameter is a string
        } // for i in range(3)
    } // if rotation_angle is not a string

    if (require_parametric_transform) {
        this->ParseAndSetParametricTransform(
            rotation_definition,
            euler_angles_parameters,
            rotation_axis_parameters,
            rotation_angle_parameters,
            reference_point_parameters,
            translation_vector_parameters);
    } // if transform is parametric
    else {
        this->ParseAndSetConstantTransform(
            rotation_definition,
            euler_angles_parameters,
            rotation_axis_parameters,
            rotation_angle_parameters,
            reference_point_parameters,
            translation_vector_parameters);
    } // transform is constant

    KRATOS_CATCH("");
}


void ImposeMeshMotionProcess::ParseAndSetConstantTransform(const std::string& rRotationDefinition,
                                                           const Parameters& rEulerAngles,
                                                           const Parameters& rRotationAxis,
                                                           const Parameters& rRotationAngle,
                                                           const Parameters& rReferencePoint,
                                                           const Parameters& rTranslationVector)
{
    KRATOS_TRY

    LinearTransform::Pointer p_transform;

    // Parse parameters - translation vector
    Vector vector_parameter = rTranslationVector.GetVector();
    if (vector_parameter.size() != 3) {
        KRATOS_ERROR << "'translation_vector' must be of size 3, but got" << vector_parameter.size();
    }
    array_1d<double,3> translation_vector;
    translation_vector[0] = vector_parameter[0];
    translation_vector[1] = vector_parameter[1];
    translation_vector[2] = vector_parameter[2];

    vector_parameter = rReferencePoint.GetVector();
    if (vector_parameter.size() != 3) {
        KRATOS_ERROR << "'reference_point' must be of size 3, but got" << vector_parameter.size();
    }
    array_1d<double,3> reference_point;
    reference_point[0] = vector_parameter[0];
    reference_point[1] = vector_parameter[1];
    reference_point[2] = vector_parameter[2];

    if (rRotationDefinition == "rotation_axis") {
        vector_parameter = rRotationAxis.GetVector();
        if (vector_parameter.size() != 3) {
            KRATOS_ERROR << "'rotation_axis' must be of size 3, but got" << vector_parameter.size();
        }

        array_1d<double,3> rotation_axis;
        rotation_axis[0] = vector_parameter[0];
        rotation_axis[1] = vector_parameter[1];
        rotation_axis[2] = vector_parameter[2];

        const double rotation_angle = rRotationAngle.GetDouble();

        p_transform = std::make_shared<LinearTransform>(
            rotation_axis,
            rotation_angle,
            reference_point,
            translation_vector);
    } // rotation_definition == "rotation_axis"

    else if (rRotationDefinition == "euler_angles") {
        vector_parameter = rEulerAngles.GetVector();
        if (vector_parameter.size() != 3) {
            KRATOS_ERROR << "'euler_angles' must be of size 3, but got" << vector_parameter.size();
        }

        array_1d<double,3> euler_angles;
        euler_angles[0] = vector_parameter[0];
        euler_angles[1] = vector_parameter[1];
        euler_angles[2] = vector_parameter[2];

        p_transform = std::make_shared<LinearTransform>(
            euler_angles,
            reference_point,
            translation_vector);
    } // rotation_definition == "euler_angles"

    else {
        KRATOS_ERROR << "Invalid 'rotation_definition': " << rRotationDefinition;
    } // unsupported rotation_definition

    // Set transform function
    mTransformFunctor = [p_transform, this](const Node<3>& rNode)
    {
        KRATOS_TRY
        return p_transform->Apply(rNode);
        KRATOS_CATCH("");
    };

    KRATOS_CATCH("");
}


void ImposeMeshMotionProcess::ParseAndSetParametricTransform(const std::string& rRotationDefinition,
                                                             const Parameters& rEulerAngles,
                                                             const Parameters& rRotationAxis,
                                                             const Parameters& rRotationAngle,
                                                             const Parameters& rReferencePoint,
                                                             const Parameters& rTranslationVector)
{
    KRATOS_TRY

    ParametricLinearTransform::Pointer p_transform;

        if (rRotationDefinition == "rotation_axis") {
            p_transform = std::make_shared<ParametricLinearTransform>(
                rRotationAxis,
                rRotationAngle,
                rReferencePoint,
                rTranslationVector);
        } // rotation_definition == "rotation_axis"

        else if (rRotationDefinition == "euler_angles") {
            p_transform = std::make_shared<ParametricLinearTransform>(
                rEulerAngles,
                rReferencePoint,
                rTranslationVector);
        } // rotation_definition == "euler_angles"

        else {
            KRATOS_ERROR << "Invalid 'rotation_definition': " << rRotationDefinition;
        } // unsupported rotation_definition

        mTransformFunctor = [p_transform, this](const Node<3>& rNode)
        {
            KRATOS_TRY

            const double time = mrModelPart.GetProcessInfo().GetValue(TIME);

            return p_transform->Apply(
                rNode,
                time,
                rNode.X0(),
                rNode.Y0(),
                rNode.Z0());

            KRATOS_CATCH("");
        };

    KRATOS_CATCH("");
}


void ImposeMeshMotionProcess::ExecuteInitializeSolutionStep()
{
    KRATOS_TRY;

    const double time = mrModelPart.GetProcessInfo().GetValue(TIME);

    if (mIntervalUtility.IsInInterval(time)) {
        block_for_each(mrModelPart.Nodes(),
            [this](Node<3>& rNode) {
                array_1d<double,3> transformed_point = mTransformFunctor(rNode);
                rNode.GetSolutionStepValue(MESH_DISPLACEMENT) = transformed_point - rNode;
            }
        ); // block_for_each
    } // if time in interval

    KRATOS_CATCH("");
}


const Parameters ImposeMeshMotionProcess::GetDefaultParameters() const
{
    return Parameters(R"(
    {
        "model_part_name"       : "",
        "interval"              : [0.0, "End"],
        "rotation_definition"   : "rotation_axis",
        "euler_angles"          : [0.0, 0.0, 0.0],
        "rotation_axis"         : [0.0, 0.0, 1.0],
        "reference_point"       : [0.0, 0.0, 0.0],
        "rotation_angle"        : 0,
        "translation_vector"    : [0.0, 0.0, 0.0]
    })");
}


std::string ImposeMeshMotionProcess::Info() const
{
    return "ImposeMeshMotionProcess";
}


void ImposeMeshMotionProcess::PrintInfo(std::ostream& rStream) const{
    rStream << this->Info();
}


std::ostream& operator<<(std::ostream& rStream,
                         ImposeMeshMotionProcess& rThis)
{
    rThis.PrintInfo(rStream);
    rStream << std::endl;
    rThis.PrintData(rStream);
    return rStream;
}


} // namespace Kratos

File: master/applications/MeshMovingApplication/python_scripts/impose_mesh_motion_process.py
# --- Core Imports ---
import KratosMultiphysics

# --- MeshMoving Imports ---
import KratosMultiphysics.MeshMovingApplication as MeshMovingApplication


def Factory(parameters, model):
    if not isinstance(parameters, KratosMultiphysics.Parameters):
        raise Exception(
            'expected input shall be a Parameters object, encapsulating a json string')
    return ImposeMeshMotionProcess(model, parameters["Parameters"])


class ImposeMeshMotionProcess(KratosMultiphysics.Process):
    """ @brief Impose a rotation followed by translation on a ModelPart.

        @details The transformation is equivalent to:
                 1) Translation to the reference frame (offset the origin)
                 2) Specified rotation
                 3) Reverse translation from the reference frame (undo origin offset)
                 4) Specified translation
                 Note: angles in radians

                 The rotation can be defined by either "euler_angles"
                 or a "rotation_axis" and "rotation_angle" pair. The following parameters can be
                 defined parametrically (see GenericFunctionUtility):
                 "euler_angles", "rotation_axis", "reference_point", "rotation_angle", "translation_vector"

                Default parameters:
                @code
                {
                    "model_part_name"       : "",
                    "interval"              : [0.0, "End"],
                    "rotation_definition"   : "rotation_axis",
                    "euler_angles"          : [0.0, 0.0, 0.0],
                    "rotation_axis"         : [0.0, 0.0, 1.0],
                    "reference_point"       : [0.0, 0.0, 0.0]
                    "rotation_angle"        : 0,
                    "translation_vector"    : [0.0, 0.0, 0.0],
                }
                @endcode

        @note the euler angles follow the convention specified by @ref{Quaternion} (Z, -X', Z")
    """

    def __init__(self,
                 model: KratosMultiphysics.Model,
                 parameters: KratosMultiphysics.Parameters):
        """Impose a rotation followed by translation on a ModelPart."""
        KratosMultiphysics.Process.__init__(self)

        # 'rotation_angle' can either be a float or "End", but the validator can't handle both
        # => convert 'rotation_angle' to the input type
        default_parameters = self.GetDefaultParameters()
        if parameters.Has("rotation_angle") and parameters["rotation_angle"].IsString():
            default_parameters.RemoveValue("rotation_angle")
            default_parameters.AddValue("rotation_angle", parameters["rotation_angle"])

        parameters.ValidateAndAssignDefaults(default_parameters)
        self.model_part = model[parameters["model_part_name"].GetString()]

        # Parse interval
        self.interval_utility = KratosMultiphysics.IntervalUtility(parameters)

        # Determine whether a constant transform will suffice or a parametric one is needed
        requires_parametric_transform = False

        euler_angle_parameters = parameters["euler_angles"]
        rotation_axis_parameters = parameters["rotation_axis"]
        rotation_angle_parameters = parameters["rotation_angle"]
        reference_point_parameters = parameters["reference_point"]
        translation_vector_parameters = parameters["translation_vector"]

        if rotation_angle_parameters.IsString():
            requires_parametric_transform = True
        else:
            for i in range(3):
                if euler_angle_parameters[i].IsString() or rotation_axis_parameters[i].IsString() or reference_point_parameters[i].IsString() or translation_vector_parameters[i].IsString():
                    requires_parametric_transform = True
                    break

        rotation_definition = parameters["rotation_definition"].GetString()

        if requires_parametric_transform:
            if rotation_definition == "rotation_axis":
                self.transform = MeshMovingApplication.ParametricAffineTransform(
                    rotation_axis_parameters,
                    rotation_angle_parameters,
                    reference_point_parameters,
                    translation_vector_parameters)
            elif rotation_definition == "euler_angles":
                self.transform = MeshMovingApplication.ParametricAffineTransform(
                    euler_angle_parameters,
                    reference_point_parameters,
                    translation_vector_parameters)
            else:
                raise ValueError("Invalid rotation definition '{}'".format(rotation_definition))

        else:
            reference_point = reference_point_parameters.GetVector()
            translation_vector = translation_vector_parameters.GetVector()
            if rotation_definition == "rotation_axis":
                rotation_axis = rotation_axis_parameters.GetVector()
                rotation_angle = rotation_angle_parameters.GetDouble()
                self.transform = MeshMovingApplication.AffineTransform(
                    rotation_axis,
                    rotation_angle,
                    reference_point,
                    translation_vector)
            elif rotation_definition == "euler_angles":
                euler_angles = euler_angle_parameters.GetVector()
                self.transform = MeshMovingApplication.AffineTransform(
                    euler_angles,
                    reference_point,
                    translation_vector)
            else:
                raise ValueError("Invalid rotation definition '{}'".format(rotation_definition))


    def ExecuteInitializeSolutionStep(self):
        time = self.model_part.ProcessInfo[KratosMultiphysics.TIME]

        if self.interval_utility.IsInInterval(time):
            MeshMovingApplication.MoveModelPart(self.model_part, self.transform)


    @staticmethod
    def GetDefaultParameters():
        return KratosMultiphysics.Parameters("""{
            "model_part_name"       : "",
            "interval"              : [0.0, "End"],
            "rotation_definition"   : "rotation_axis",
            "euler_angles"          : [0.0, 0.0, 0.0],
            "rotation_axis"         : [0.0, 0.0, 1.0],
            "reference_point"       : [0.0, 0.0, 0.0],
            "rotation_angle"        : 0,
            "translation_vector"    : [0.0, 0.0, 0.0]
        }""")